### PR TITLE
Feature/speakers emails

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=2948118&fileview=grid

ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=2948603&fileview=grid

ref https://tipit.avaza.com/project/view/304272#!tab=task-pane&task=2962049
Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php [hunk 1]
```diff
@@ -177,6 +177,42 @@ public function serialize($expand = null, array $fields = [], array $relations =
...
+                            $accepted_presentations[] = SerializerRegistry::getInstance()->getSerializer($p)->serialize(AbstractSerializer::filterExpandByPrefix($expand, $relation));
+                        }
+                        $moderated_accepted_presentations = [];
+                        foreach ($speaker->getPublishedRegularPresentations($summit, PresentationSpeaker::RoleModerator, true) as $p) {
+                            $moderated_accepted_presentations[] = SerializerRegistry::getInstance()->getSerializer($p)->serialize(AbstractSerializer::filterExpandByPrefix($expand, $relation));
+                        }
+                        $values['accepted_presentations'] = array_merge($accepted_presentations, $moderated_accepted_presentations);
+                    }
+                    break;
+                    case 'alternate_presentations': {
```
Comment (smarcet) [inline app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php:192]: need to add this on allowed_relations and pass the ids on a non expand

File: app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php [hunk 2]
```diff
@@ -177,6 +177,42 @@ public function serialize($expand = null, array $fields = [], array $relations =
...
+                            $alternate_presentations[] = SerializerRegistry::getInstance()->getSerializer($p)->serialize(AbstractSerializer::filterExpandByPrefix($expand, $relation));
+                        }
+                        $moderated_alternate_presentations = [];
+                        foreach ($speaker->getAlternatePresentations($summit, PresentationSpeaker::ROLE_MODERATOR) as $p) {
+                            $moderated_alternate_presentations[] = SerializerRegistry::getInstance()->getSerializer($p)->serialize(AbstractSerializer::filterExpandByPrefix($expand, $relation));
+                        }
+                        $values['alternate_presentations'] = array_merge($alternate_presentations, $moderated_alternate_presentations);
+                    }
+                    break;
+                    case 'rejected_presentations': {
```
Comment (smarcet) [inline app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php:204]: need to add this on allowed_relations and pass the ids on a non expand

File: app/Repositories/Summit/DoctrineSpeakerRepository.php [hunk 3]
```diff
@@ -30,6 +30,53 @@ final class DoctrineSpeakerRepository
...
+                                     WHEN CustomOrder <= SessionCount THEN 'accepted'
+                                     ELSE 'alternate'
+                                 END AS SelectionStatus
+                          FROM (
+                                   SELECT p2.CustomOrder AS CustomOrder, pc.SessionCount AS SessionCount
+                                   FROM Presentation p2
+                                            LEFT JOIN SummitSelectedPresentation sp ON sp.PresentationID = p2.ID
+                                            LEFT JOIN SummitSelectedPresentationList l ON sp.SummitSelectedPresentationListID = l.ID
+                                            LEFT JOIN PresentationCategory pc on l.CategoryID = pc.ID
+                                   WHERE p2.ID = P.ID AND sp.Collection = 'selected'
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrineSpeakerRepository.php:70]: there are missing where condition on list
check https://github.com/OpenStackweb/summit-api/blob/57b49bc0164c9e9b3ef29f3132cd5f6beadbf72d/app/Models/Foundation/Summit/Events/Presentations/Presentation.php#L833

File: app/Repositories/Summit/DoctrineSpeakerRepository.php [hunk 4]
```diff
@@ -30,6 +30,53 @@ final class DoctrineSpeakerRepository
...
+                                   UNION
+                                   SELECT null, null
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrineSpeakerRepository.php:72]: why is this ?

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments
regards

File: app/Repositories/Summit/DoctrineSpeakerRepository.php (hunk 4 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSpeakerRepository.php:72]: It's a fallback. If no rows are returned from the first part of the union, the second part will act as a dummy to fulfill the CASE WHEN -> 'unaccepted' condition

File: app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php [hunk 5]
```diff
@@ -177,6 +177,42 @@ public function serialize($expand = null, array $fields = [], array $relations =
                         $values['moderated_presentations'] = $moderated_presentations;
                     }
                     break;
+                    case 'accepted_presentations': {
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php:201]: done

File: app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php (hunk 2 above)
Comment (romanetar) [reply app/ModelSerializers/Summit/Speakers/PresentationSpeakerSerializer.php:225]: done

File: app/Repositories/Summit/DoctrineSpeakerRepository.php (hunk 4 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSpeakerRepository.php]: subquery adjusted and code comments added

**Comment by romanetar** [comment]
@smarcet  PR amended as discussed

File: app/Jobs/Emails/PresentationSubmissions/ProcessSpeakersEmailRequestJob.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        (
+            sprintf
+            (
+                "ProcessSpeakersEmailRequestJob::handle summit id %s payload %s",
+                $this->summit->getId(),
+                json_encode($this->payload)
+            )
+        );
+
+        $filter = !is_null($this->filter) ? FilterParser::parse($this->filter, [
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/ProcessSpeakersEmailRequestJob.php:65]: filter needs to match the get counter part

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedAlternateEmail.php [hunk 7]
```diff
@@ -35,27 +39,41 @@ protected function getEmailEventSlug(): string
...
     (
         Summit $summit,
         SummitRegistrationPromoCode $promo_code,
         PresentationSpeaker $speaker,
-        string $speaker_role,
         string $confirmation_token
     ){
         parent::__construct($summit, $speaker, $promo_code);
 
         $summit = $promo_code->getSummit();
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedAlternateEmail.php:53]: this line is not needed , summit is already a param

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedOnlyEmail.php [hunk 8]
```diff
@@ -35,23 +38,30 @@ protected function getEmailEventSlug(): string
...
         Summit $summit,
         SummitRegistrationPromoCode $promo_code,
         PresentationSpeaker $speaker,
-        string $speaker_role,
         string $confirmation_token
     )
     {
         parent::__construct($summit, $speaker, $promo_code);
 
         $summit = $promo_code->getSummit();
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedOnlyEmail.php:53]: this line is not needed summit is already a param

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedRejectedEmail.php [hunk 9]
```diff
@@ -35,28 +38,42 @@ protected function getEmailEventSlug(): string
...
         Summit $summit,
         SummitRegistrationPromoCode $promo_code,
         PresentationSpeaker $speaker,
-        string $speaker_role,
         string $confirmation_token
     )
     {
         parent::__construct($summit, $speaker, $promo_code);
 
         $summit = $promo_code->getSummit();
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedRejectedEmail.php:53]: this line is not needed

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateRejectedEmail.php [hunk 10]
```diff
@@ -35,28 +38,42 @@ protected function getEmailEventSlug(): string
...
         Summit $summit,
         SummitRegistrationPromoCode $promo_code,
         PresentationSpeaker $speaker,
-        string $speaker_role,
         string $confirmation_token
     )
     {
         parent::__construct($summit, $speaker, $promo_code);
 
         $summit = $promo_code->getSummit();
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateRejectedEmail.php:53]: not needed

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessEmail.php [hunk 11]
```diff
@@ -42,6 +43,13 @@ public function __construct
         $payload['summit_site_url'] = $summit->getDefaultPageUrl();
         $payload['speaker_full_name'] = $speaker->getFullName();
         $payload['speaker_email'] = $speaker->getEmail();
+
+        $test_email_recipient = EmailTest::getEmailAddress();
+
+        if (!is_null($test_email_recipient)) {
```
Comment (smarcet) [inline app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessEmail.php:49]: isnt better to test by empty ( that includes null ) ?

File: app/Services/Utils/Facades/EmailTest.php [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ **/
+
+use Illuminate\Support\Facades\Facade;
+
+/**
+ * Class EmailTest
+ * @package services\utils\Facades
+ */
+class EmailTest extends Facade
```
Comment (smarcet) [inline app/Services/Utils/Facades/EmailTest.php:21]: where are u binding this facade to container
please check this example
https://github.com/OpenStackweb/summit-api/blob/ade98172ad7563b781cbaa75eda06d5fd03d4d87/app/Providers/AppServiceProvider.php#L631

File: app/Services/Utils/Facades/EmailExcerpt.php [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ **/
+
+use Illuminate\Support\Facades\Facade;
+
+/**
+ * Class EmailExcerpt
+ * @package services\utils\Facades
+ */
+class EmailExcerpt extends Facade
```
Comment (smarcet) [inline app/Services/Utils/Facades/EmailExcerpt.php:21]: where are u binding this facade to container ?
check https://github.com/OpenStackweb/summit-api/blob/ade98172ad7563b781cbaa75eda06d5fd03d4d87/app/Providers/AppServiceProvider.php#L631

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments
regards

File: app/Services/Utils/IEmailExcerptService.php [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ **/
+
+/**
+ * Class EmailExcerptService
+ * @package App\Services\utils
+ */
+interface IEmailExcerptService
+{
+    public function add($value);
```
Comment (smarcet) [inline app/Services/Utils/IEmailExcerptService.php:22]: could we define the types ?
$value is mixed or a string ?

File: app/Services/Utils/IEmailExcerptService.php [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+    public function clearReport();
```
Comment (smarcet) [inline app/Services/Utils/IEmailExcerptService.php:24]: add :void

File: app/Services/Utils/IEmailExcerptService.php [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+/**
+ * Class EmailExcerptService
+ * @package App\Services\utils
+ */
+interface IEmailExcerptService
+{
```
Comment (smarcet) [inline app/Services/Utils/IEmailExcerptService.php:21]: please add php doc comments

File: app/Services/Utils/EmailExcerptService.php [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ **/
+
+/**
+ * Class EmailExcerptService
+ * @package App\Services\utils
+ */
+final class EmailExcerptService implements IEmailExcerptService
+{
+    private $report = [];
```
Comment (smarcet) [inline app/Services/Utils/EmailExcerptService.php:22]: please add type definitions and php docs comments

File: app/Repositories/Summit/DoctrineSpeakerRepository.php [hunk 18]
```diff
@@ -30,6 +32,84 @@ final class DoctrineSpeakerRepository
     extends SilverStripeDoctrineRepository
     implements ISpeakerRepository
 {
+    private function buildHasPresentationSubQuery(Filter $filter)
```
Comment (smarcet) [inline app/Repositories/Summit/DoctrineSpeakerRepository.php:35]: please add php docs comments / definitions

File: app/Services/Utils/Facades/EmailTest.php (hunk 12 above)
Comment (romanetar) [reply app/Services/Utils/Facades/EmailTest.php:21]: marked as resolved as was discussed

File: app/Jobs/Emails/PresentationSubmissions/ProcessSpeakersEmailRequestJob.php (hunk 6 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/ProcessSpeakersEmailRequestJob.php:65]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedAlternateEmail.php (hunk 7 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedAlternateEmail.php:53]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedOnlyEmail.php (hunk 8 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedOnlyEmail.php:53]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedRejectedEmail.php (hunk 9 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAcceptedRejectedEmail.php:53]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateOnlyEmail.php [hunk 19]
```diff
@@ -35,24 +38,30 @@ protected function getEmailEventSlug(): string
...
         Summit $summit,
         SummitRegistrationPromoCode $promo_code,
         PresentationSpeaker $speaker,
-        string $speaker_role,
         string $confirmation_token
     )
     {
         parent::__construct($summit, $speaker, $promo_code);
 
         $summit = $promo_code->getSummit();
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateOnlyEmail.php:53]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateRejectedEmail.php (hunk 10 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessAlternateRejectedEmail.php:53]: done

File: app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessEmail.php (hunk 11 above)
Comment (romanetar) [reply app/Jobs/Emails/PresentationSubmissions/SelectionProcess/PresentationSpeakerSelectionProcessEmail.php:49]: done

File: app/Services/Utils/IEmailExcerptService.php (hunk 15 above)
Comment (romanetar) [reply app/Services/Utils/IEmailExcerptService.php:24]: done

File: app/Services/Model/ISpeakerService.php [hunk 20]
```diff
@@ -193,4 +195,18 @@ public function addSpeakerBigPhoto($speaker_id, UploadedFile $file,  $max_file_s
      * @throws EntityNotFoundException
      */
     public function deleteSpeakerBigPhoto($speaker_id):void;
+
+    /**
+     * @param Summit $summit
+     * @param array $payload
+     * @param mixed $filter
+     */
+    public function triggerSend(Summit $summit, array $payload, $filter = null):void;
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Services/Model/ISpeakerService.php:204]: done

File: app/Services/Utils/EmailTestDTO.php [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ **/
+
+/**
+ * Class EmailTestDTO
+ * @package App\Services\utils
+ */
+final class EmailTestDTO
+{
+    private $email_address;
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Services/Utils/EmailTestDTO.php:22]: done

File: app/Services/Utils/EmailExcerptService.php (hunk 17 above)
Comment (romanetar) [reply app/Services/Utils/EmailExcerptService.php:22]: done

File: app/Repositories/Summit/DoctrineSpeakerRepository.php (hunk 18 above)
Comment (romanetar) [reply app/Repositories/Summit/DoctrineSpeakerRepository.php:35]: done

File: app/Services/Model/Strategies/EmailActions/SpeakerActionsEmailStrategy.php [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+                    break;
+                case PresentationSpeakerSelectionProcessRejectedEmail::EVENT_SLUG:
+                    if (!$has_accepted_presentations && !$has_alternate_presentations && $has_rejected_presentations) {
+                        $type = SpeakerAnnouncementSummitEmail::TypeRejected;
+                    }
+                    break;
+                default:
+                    return null;
+            }
+
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Services/Model/Strategies/EmailActions/SpeakerActionsEmailStrategy.php:119]: done

**Comment by romanetar** [comment]
@smarcet PR amended solving all open issues

**Pull request merged**